ocaml/typing/primitive.ml

```diff
@@ -380,11 +402,157 @@ let native_name_is_external p =
   let nat_name = native_name p in
   nat_name <> "" && nat_name.[0] <> '%'
 
-let sort_of_native_repr = function
-  | Same_as_ocaml_repr s -> s
-  | (Unboxed_float _ | Unboxed_integer _ | Untagged_int |
-     Unboxed_vector _) ->
-    Jkind.Sort.Value
+module Repr_check = struct
+
+  type result =
+    | Wrong_arity
+    | Wrong_repr
+    | Success
+
+  let args_res_reprs prim =
+    (prim.prim_native_repr_args @ [prim.prim_native_repr_res])
+    |> List.map snd
+
+  let is repr = equal_native_repr repr
+
+  let any = fun _ -> true
+
+  let value_or_unboxed_or_untagged = function
+    | Same_as_ocaml_repr Value
+    | Unboxed_float _ | Unboxed_integer _ | Unboxed_vector _
+    | Untagged_int -> true
+    | Same_as_ocaml_repr _ | Repr_poly -> false
+
+  let check checks prim =
+    let reprs = args_res_reprs prim in
+    if List.length reprs <> List.length checks
+    then Wrong_arity
+    else
+    if not (List.for_all2 (fun f x -> f x) checks reprs)
+    then Wrong_repr
+    else Success
+
+  let exactly required =
+    check (List.map is required)
+
+  let same_arg_res_repr_with_arity arity prim =
+    let repr = snd prim.prim_native_repr_res in
+    exactly
+      (List.init (arity+1) (fun _ -> repr))
+      prim
+
+  let no_non_value_repr prim =
+    let arity = List.length prim.prim_native_repr_args in
+    check
+      (List.init (arity+1) (fun _ -> value_or_unboxed_or_untagged))
+      prim
+end
+
+let prim_has_valid_reprs ~loc prim =
```

Comment:
Added a note clarifying that "any" in this context refers to any value of type `native_repr` instead of jkind `any`:

https://github.com/ocaml-flambda/flambda-backend/blob/4470b558d62f2a457ffda77d2cca79406d30103d/ocaml/typing/primitive.ml#L451-L456